internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go

```diff
@@ -505,6 +505,147 @@ func (s *Server) List(req *machine.ListRequest, obj machine.MachineService_ListS
 	return nil
 }
 
+// DiskUsage implements the machine.MachineServer interface.
+func (s *Server) DiskUsage(req *machine.DiskUsageRequest, obj machine.MachineService_DiskUsageServer) error { //nolint: gocyclo
+	if req == nil {
+		req = new(machine.DiskUsageRequest)
+	}
+
+	for _, path := range req.Paths {
+		if !strings.HasPrefix(path, OSPathSeparator) {
+			// Make sure we use complete paths
+			path = OSPathSeparator + path
+		}
+
+		path = strings.TrimSuffix(path, OSPathSeparator)
+		if path == "" {
+			path = "/"
+		}
+
+		_, err := os.Stat(path)
+		if err == os.ErrNotExist {
+			return fmt.Errorf("%s no such file or directory", path)
+		}
+
+		files, err := archiver.Walker(obj.Context(), path, archiver.WithMaxRecurseDepth(-1))
+		if err != nil {
+			return err
+		}
+
+		folders := map[string]*machine.DiskUsageInfo{}
+
+		// send a record back to client if the message shouldn't be skipped
+		// at the same time use record information for folder size estimation
+		sendSize := func(info *machine.DiskUsageInfo, depth int32, isDir bool) {
+			prefix := strings.TrimRight(filepath.Dir(info.Name), "/")
+			if folder, ok := folders[prefix]; ok {
```

Comment:
should we increase all the folder sizes walking "up" the filesystem hiearachy here? 

(unless I missed something, I don't see how `/bin` size gets incremented for `/bin/directory/file` ?

Reply:
it should work.
It should report it like that for example

`100 /bin/directory/file` -> add 100 to `/bin/directory` size.
`4196 /bin/directory` -> add 4196 to `/bin` size.

Reply:
ahh, got it, it will report subdirectory overall size and add it back to the parent directory